Lib/test/test_asyncio/test_sendfile.py

```diff
@@ -446,6 +446,8 @@ def test_sendfile_ssl_close_peer_after_receiving(self):
         self.assertEqual(srv_proto.data, self.DATA)
         self.assertEqual(self.file.tell(), len(self.DATA))
 
+    @unittest.skipIf(sys.platform.startswith('sunos'),
+                     "Doesn't work on Solaris")
     def test_sendfile_close_peer_in_the_middle_of_receiving(self):
```

Comment:
This single test fails with "_AssertionError: ConnectionError not raised_" and I am not sure why. With this patch, sendfile is not called if an offset is too high, which is slightly different from other platforms where it is called and sends nothing; but I don't think that causes this issue.

It seems to me that the entire message is sent and `close_after` doesn't work as expected. But I don't see into asyncio internals that much, so it's just a guess.

It was failing before as well, so it doesn't seem to be a regression. If you have any idea why that might be the case, how to test it further, or how to fix that, I will gladly do so (I don't really want to skip it).

Reply:
> It was failing before as well, so it doesn't seem to be a regression. If you have any idea why that might be the case, how to test it further, or how to fix that, I will gladly do so (I don't really want to skip it).

Yeah, I don't like skipping it either. Unfortunately I won't be able to assist with the debugging, but if you have time please try to figure this out.

Reply:
I am getting closer. It looks like no matter the `BUF_SIZE`, `sendfile` always sends at least `131072` bytes, so it sends everything before transport is closed and hence no `ConnectionError` is raised. If I understand it correctly, the intended behavior is to limit `sendfile` to send at maximum `BUF_SIZE` bytes, meaning that transport is closed long before everything is sent (correct me if I'm wrong). I found out that the test passes if `DATA` size is doubled.

My guess is that this might be a difference in how Solaris handles those buffer limits; I will keep digging.

Reply:
I believe I found the culprit - the reason is that [as per documentation](https://docs.oracle.com/cd/E88353_01/html/E37843/setsockopt-3c.html): _"At present, lowering SO_RCVBUF on a TCP connection after it has been established has no effect"_. This is on Oracle Solaris, but the bug is old enough that it should be present in all OpenSolaris forks unless they fixed it themselves.

I don't know If you'd prefer having it skipped here for all Solarises or if we should skip it internally since it is an OS bug, not a difference (I think that both are valid).

I checked Illumos source code and I believe they are also affected by this `SO_RCVBUF` related bug (but I might have missed their fix somewhere where I was not looking and I cannot verify that on Illumos machine).

Reply:
Let's skip the test on Solaris and add an elaborate comment why, basically summarizing

> "At present, lowering SO_RCVBUF on a TCP connection after it has been established has no effect". This is on Oracle Solaris, but the bug is old enough that it should be present in all OpenSolaris forks unless they fixed it themselves.